### ISSUE TITLE
Render throttling

### ISSUE THREAD

**Issue by theKashey** [airbnb/enzyme#1418]
Nowadays more and more people start using enzyme for integration tests, abandoning the lands of Webdriver.
Enzyme is a great thing for unit tests, but not the perfect for the integration ones. 

And that's why: React throttles `setState`, applying the changes on next tick. React doing the same for `render` materializing VDom not instantly.
As result, if on componentWillMount you will dispatch redux event, you will get the old store during the render.

Current synchronius model is ideal for units tests, but is it possible to make enzyme behaviour more React-ish? Add some debounce to make it more async, and more real?

**Comment by matthew-matvei** [comment]
@theKashey just an idea, and it might only be a current round-about way, but have you looked into [Sinon's fake timers](http://sinonjs.org/releases/v4.1.3/fake-timers/)? Perhaps there's a nice way to fake advancing the clock forward to trigger React's `setState` call (completely untried, just a thought)

**Comment by ljharb** [comment]
In what way is enzyme’s behavior not currently the same as React’s?

If React throttles setState and we don’t, that’s a bug and we should fix it.

**Comment by theKashey** [comment]
@ljharb - the current "synchronous" way is much more unit test friendly, especially keeping in mind `shallow`. Or `mount` with `faked` environment.
You have to keep it.

@matthew-matvei - in this case, I have to advance clock backwards.

**Comment by ljharb** [comment]
I don't agree we'd have to keep it; we could certainly make `shallow` or `mount` provide a Promise for a wrapper.

We could implement it as an option in a semver-minor, and if it ended up being useful, we could flip the default in the next semver-major.

**Comment by theKashey** [comment]
@ljharb - you are the boss here. 
Anyway - the "server-side" React had the same synchronous behaviour as the node-js powered enzyme.
As long ezyme could be used to test both front-end and back-end - both ways should coexist.

**Comment by ljharb** [comment]
I think a PR that adds an option to enable the throttled behavior you expect, would be the best way to proceed here.
